### PR TITLE
[Validator] String normalization options for string-based validators

### PR DESCRIPTION
**Pull request by renan-taranto**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #26239
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/9433

Todo:
- [x] Document the new options
- [x] Update Doc PR

Add trimming options to the string constraints.

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/NotBlankValidator.php [hunk 1]
```diff
@@ -29,7 +29,9 @@ public function validate($value, Constraint $constraint)
             throw new UnexpectedTypeException($constraint, __NAMESPACE__.'\NotBlank');
         }
 
-        if (false === $value || (empty($value) && '0' != $value)) {
+        $normalizedValue = $constraint->trim ? trim($value) : $value;
```
Comment (Destroy666x) [inline src/Symfony/Component/Validator/Constraints/NotBlankValidator.php]: You need to check whether the value is a string in this validator - someone may use the option incorrectly and `trim()` expects it.

Also, would be nice if you added tests for Unicode whitespaces, e.g. `\u{00A0} test` - I'm pretty sure that in PHP 7.1.x both `trim()` etc. and `\s` regex meta character don't consider them still. See: https://stackoverflow.com/questions/4166896/trim-unicode-whitespace-in-php-5-2

Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/NotBlankValidator.php]: Thanks @Destroy666x. Now the regex also works for every Unicode whitespace.

**Review by Destroy666x — APPROVED** [review]

**Comment by Destroy666x** [comment]
Looks good now, thanks. Some more edge cases could be added for tests, but not sure how strict are they here yet

**Comment by renan-taranto** [comment]
Thanks for the review @Destroy666x. Do you mean more edge cases for the regexps on all of the unicode whitespaces?

File: src/Symfony/Component/Validator/Constraints/LengthValidator.php [hunk 2]
```diff
@@ -38,9 +38,15 @@ public function validate($value, Constraint $constraint)
         }
 
         $stringValue = (string) $value;
+        $normalizedString = $stringValue;
+
+        $normalizedString = $constraint->ltrim ? preg_replace('/^[\pZ\pC]+/u', '', $normalizedString) : $normalizedString;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/LengthValidator.php]: This makes a strong supposition, namely that the input is in UTF-8. Is that guaranteed to be the case?

Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/LengthValidator.php]: Hi @nicolas-grekas you are right, thanks.
Do you believe that converting to ``UTF-8`` using the already existing ``charset`` option is a good approach (as the code below suggests)? Do you have any other/better idea?
```php
$stringValue = (string) $value;

if (!$invalidCharset = !@mb_check_encoding($stringValue, $constraint->charset)) {
    $normalizedString = mb_convert_encoding($stringValue, 'UTF-8', $constraint->charset);

    $normalizedString = $constraint->ltrim ? preg_replace('/^[\pZ\pC]+/u', '', $normalizedString) : $normalizedString;
    $normalizedString = $constraint->rtrim ? preg_replace('/[\pZ\pC]+$/u', '', $normalizedString) : $normalizedString;

    $normalizedString = $constraint->mergeConsecutiveWhitespaces ? preg_replace('/[\pZ\pC]+/u', ' ', $normalizedString) : $normalizedString;

    $length = mb_strlen($normalizedString, 'UTF-8');
}
```

File: src/Symfony/Component/Validator/Constraints/Length.php [hunk 3]
```diff
@@ -39,6 +39,9 @@ class Length extends Constraint
     public $max;
     public $min;
     public $charset = 'UTF-8';
+    public $ltrim;
```
Comment (ostrolucky) [inline src/Symfony/Component/Validator/Constraints/Length.php]: let's have some docblocks please

File: src/Symfony/Component/Validator/Constraints/LengthValidator.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/LengthValidator.php]: That's too heavy for a constraint IMHO.
I'd suggest sticking to regular spaces (ie use `ltrim` and a non unicode regex)
I'm also personally not convinced by the `mergeConsecutiveWhitespaces`, but maybe others see the use case

File: src/Symfony/Component/Validator/Constraints/Length.php (hunk 3 above)
Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/Length.php]: Right now there's no Constraint classes with docblocks, but I will work on another PR to add it in all of those.

File: src/Symfony/Component/Validator/Constraints/LengthValidator.php (hunk 2 above)
Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/LengthValidator.php]: Yeah, it also looks error prone somehow. Besides, PHP's ```trim``` functions behavior are well known and we should probably stick to it.
I will leave the ```mergeConsecutiveWhitespaces``` for another PR.

Comment (Destroy666x) [reply src/Symfony/Component/Validator/Constraints/LengthValidator.php]: @nicolas-grekas @renan-taranto I mentioned an use case in my issue - e.g. whitespace-styled forum posts that need to have enough content. 

As for "too heavy" - IMO validators are supposed to cover as many cases as possible. Unicode whitespaces are not uncommon these days. Not sure why you wouldn't cover them when it's possible to cover them.

Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/LengthValidator.php]: @Destroy666x I believe that someone using the ```mergeConsecutiveWhitespaces``` will have to implement a regex on the value (eg.: for persisting) after validating and it may yield different results than the validator, that would require knowledge about the constraint code and that's not good. PHP's ```trim``` functions are well known and wont cause that kind of confusion, for that same reason we should probably use them for the  ```ltrim``` and ```rtrim``` options.

File: src/Symfony/Component/Validator/Constraints/Length.php (hunk 3 above)
Comment (fabpot) [reply src/Symfony/Component/Validator/Constraints/Length.php]: We don't add docblocks on properties when they are strings (and here, names are pretty explicit anyway).

Comment (ostrolucky) [reply src/Symfony/Component/Validator/Constraints/Length.php]: This is boolean though. Name isn't explicit enough here. This could be also integer (number of chars to trim) or string (which chars to trim). But if that's your policy to not add docblocks here, ok then..

**Comment by xabbuh** [comment]
Would it make sense to update other constraint validators too?

**Comment by renan-taranto** [comment]
Hey guys, how do you like this PR right now?

**Review by nicolas-grekas — APPROVED** [review]

**Comment by xabbuh** [comment]
@renan-taranto For consistency, I would add this option to the [other string constraints](http://symfony.com/doc/current/reference/constraints.html#string-constraints) too.

**Comment by renan-taranto** [comment]
@xabbuh Done it.

**Review by fabpot — APPROVED** [review]

**Comment by xabbuh** [comment]
I wonder if two options are that handy when you want to trim on both sides which I could imagine happens more often than only trimming on one end. In Twig, the `trim` filter allows to pass a kind of "strategy". Would that be an option here too? Or does it make sense to add a third `trim` option which would enable`ltrim` and `rtrim`?

**Comment by nicolas-grekas** [comment]
Following @xabbuh's comment, what about having a single "trim" option, which would accept `true` to trim both sides, and `left` or `right` for one-side trims?

**Comment by renan-taranto** [comment]
Yes, a single trim option with 3 modes would be better for sure.
These constraints constructors will need to validate the `trim` parameter value and since the allowed values are the same for all, we would end up with a small duplication. Should we avoid that somehow? With a trait maybe?

```php
trait TrimModesTrait
{
    public static $disabledMode = false;
    public static $lTrimMode = 'left';
    public static $rTrimMode = 'right';
    public static $trimMode = true;
    
    public static function allowedTrimModes()
    {
        return array(
            self::$disabledMode,
            self::$lTrimMode,
            self::$rTrimMode,
            self::$trimMode
        );
    }
    
    private function throwExceptionIfTrimModeIsInvalid($options = null)
    {
        if (is_array($options) && array_key_exists('trim', $options) && !in_array($options['trim'], $this->allowedTrimModes(), true)) {
            throw new \InvalidArgumentException('The "trim" parameter value is not valid.');
        }
    }
}
```
And while validating
```php
$value = $constraint->trim === Length::$lTrimMode ? ltrim($value) : $value;
```
What do you think?

**Comment by ostrolucky** [comment]
How about `$normalizationCallable`<sup>*</sup> option instead. Implementation is really simple
```php
if (null !== $constraint->normalizationCallable) {
  $value = call_user_func($constraint->normalizationCallable, $value);
}
```

and usage as well

```php
$this->validator->validate($email, new Email(['normalizationCallable' => 'ltrim']));
```

- much more versatile
- simpler code
- less maintainance requirements (since it's more versatile, it's much less probable current code will need to be extended later, unlike the current proposal)
- simpler usage
- easier to document
- more SOLID (action is injected by user, instead of hardcoded)

<sub>* proposed name is subject to change</sub>

**Comment by nicolas-grekas** [comment]
> Length::$lTrimMode

honestly, I find this ugly :) The duplication is not an issue to me. We could throw an exception if a not supported value is passed.

@ostrolucky's idea is also interesting, although very different than the current one. WDYT?

**Comment by renan-taranto** [comment]
@nicolas-grekas I also prefer the small duplication, it would be cleaner.
@ostrolucky's idea is really nice and the advantages were already stated by him. Someone could argue that it is less intuitive for the user and less beginner friendly.
I vote for the `$normalizationCallable`.

**Comment by ogizanagi** [comment]
I vote for @ostrolucky 's idea too. While it may be considered less intuitive, this main use-case could be highlighted in documentation.

**Comment by xabbuh** [comment]
I very much like this proposal. We would just need to find a good name for the option. `normalizationCallable` does describe what's happening, but it seems a bit verbose. What about just `normalizer`?

**Comment by renan-taranto** [comment]
`normalizer` sounds good, its intuitive. Will we allow other callables besides `trim` methods to be passed as parameter?

**Comment by xabbuh** [comment]
I think what callable to pass is up to the user as long as it receives the string to be validated as the input and returns the normalised string. But the trim examples is a good one to explain the usage of this option in the docs.

**Comment by renan-taranto** [comment]
Hey guys, the patch is updated with the `normalizer` option.
One of the Travis-ci builds broke while running the `Symfony\Component\Process\Tests\ProcessTest::testMustRun` test, not sure why.

File: src/Symfony/Component/Validator/Constraints/NotBlankValidator.php [hunk 4]
```diff
@@ -29,6 +29,10 @@ public function validate($value, Constraint $constraint)
             throw new UnexpectedTypeException($constraint, __NAMESPACE__.'\NotBlank');
         }
 
+        if (is_string($value) && null !== $constraint->normalizer) {
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/NotBlankValidator.php]: I would move this after the `if` below. There's no need to apply the normalizer when the input is empty IMO.

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/NotBlankValidator.php]: never mind, the result of the normalizer call could lead to an empty string too

**Review by xabbuh — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Comment by Destroy666x** [comment]
I was so busy that I forgot about this - callable is a great idea, thanks for suggesting it @ostrolucky

**Review by Destroy666x — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/Email.php [hunk 5]
```diff
@@ -55,6 +55,7 @@ class Email extends Constraint
      */
     public $strict;
     public $mode;
+    public $normalizer;
```
Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/Email.php]: Don't we need to at least check that the `$normalizer` is indeed a callable as we are already validating the other options?

Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/Email.php]: Hi @fabpot. I've added the checks.

File: src/Symfony/Component/Validator/Constraints/Email.php [hunk 6]
```diff
@@ -67,5 +68,9 @@ public function __construct($options = null)
         }
 
         parent::__construct($options);
+
+        if (null !== $this->normalizer && !\is_callable($options['normalizer'])) {
+            throw new \InvalidArgumentException('The "normalizer" parameter value is not a valid callable.');
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/Email.php]: I would reword this a bit:

```php
if (null !== $this->normalizer && !\is_callable($options['normalizer'])) {
    throw new \InvalidArgumentException(sprintf('The "normalizer" option must be a valid callable, "%s" given.', is_object($this->normalizer) ? get_class($this->normalizer) : gettype($this->normalizer)));
}
```

Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/Email.php]: Sure

File: src/Symfony/Component/Validator/Constraints/Email.php [hunk 7]
```diff
@@ -74,5 +75,9 @@ public function __construct($options = null)
             // throw new LogicException(sprintf('The "egulias/email-validator" component is required to use the "%s" constraint.', __CLASS__));
             @trigger_error(sprintf('Using the "%s" constraint without the "egulias/email-validator" component installed is deprecated since Symfony 4.2.', __CLASS__), E_USER_DEPRECATED);
         }
+
+        if (null !== $this->normalizer && !\is_callable($options['normalizer'])) {
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/Email.php]: I think we could be consistent here in accessing either the property or the array element. Could be less confusing.

File: src/Symfony/Component/Validator/Constraints/Ip.php [hunk 8]
```diff
@@ -82,5 +84,9 @@ public function __construct($options = null)
         if (!\in_array($this->version, self::$versions)) {
             throw new ConstraintDefinitionException(sprintf('The option "version" must be one of "%s"', implode('", "', self::$versions)));
         }
+
+        if (null !== $this->normalizer && !\is_callable($options['normalizer'])) {
+            throw new \InvalidArgumentException(sprintf('The "normalizer" option must be a valid callable ("%s" given).', \is_object($this->normalizer) ? \get_class($this->normalizer) : \gettype($this->normalizer)));
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/Ip.php]: Don't we also have an `InvalidArgumentException` in the `Exception` namespace of the Validator component?

**Comment by gmponos** [comment]
Hello,

for me as a developer using symfony the `normalizer` option is a little bit confusing. I am referring to the naming of it. Maybe it's because in the syfmony world when I hear `normalizer` my mind goes to the `Serializer` component. So in the end when I saw the title of this PR I was expecting to see about an option that will be able to convert objects into strings and remove this code:

``` php
       if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
```

I am not sure if the same confusion will be caused later for whatever reason. Therefore my suggestion would be `formatter`. Or simple `trim` and allowing this option to be either true or callable and if it is true just use plain trim. It's just a suggestion your call :)

**Comment by ostrolucky** [comment]
You can pass even callable which makes input pass the above check, it's your choice. Hence formatter is not good name.

**Review by ostrolucky — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/Email.php (hunk 7 above)
Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/Email.php]: Good catch! Thanks.

**Comment by renan-taranto** [comment]
PR updated. AppVeyor broke due to the unrelated `Symfony\Component\Process\Tests\ProcessTest`.

**Review by nicolas-grekas — COMMENTED** [review]
Could you please and a line in the changelog of the component?

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 9]
```diff
@@ -81,6 +81,10 @@ public function validate($value, Constraint $constraint)
 
         $value = (string) $value;
 
+        if (null !== $constraint->normalizer) {
+            $value = \call_user_func($constraint->normalizer, $value);
```
[the remark this answers is not in the document]
Comment (renan-taranto) [reply src/Symfony/Component/Validator/Constraints/EmailValidator.php]: Sure. I added a line in the 4.3.0 section of the changelog, is that correct? Or it should be in the 4.2.0 section?

File: src/Symfony/Component/Validator/Constraints/Email.php [hunk 10]
```diff
@@ -98,5 +100,9 @@ public function __construct($options = null)
             // throw new LogicException(sprintf('The "egulias/email-validator" component is required to use the "%s" constraint in strict mode.', __CLASS__));
             @trigger_error(sprintf('Using the "%s" constraint in strict mode without the "egulias/email-validator" component installed is deprecated since Symfony 4.2.', __CLASS__), E_USER_DEPRECATED);
         }
+
+        if (null !== $this->normalizer && !\is_callable($this->normalizer)) {
+            throw new InvalidArgumentException(sprintf('The "normalizer" option must be a valid callable ("%s" given).', \is_object($this->normalizer) ? \get_class($this->normalizer) : \gettype($this->normalizer)));
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/Email.php]: That would introduce a new type of exception that people could catch. I'm not sure that makes sense - duplication here is legit to me.

**Comment by renan-taranto** [comment]
Hi @nicolas-grekas, is there anything else I can do for this PR?

**Review by vudaltsov — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @renan-taranto.
